server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java

```diff
@@ -136,40 +137,75 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
-                    break;
+            scoreDocs.sort(Comparator.comparingInt(scoreDoc -> scoreDoc.doc));
+            // It is possible that the different results refer to the same doc.
+            for (int i = 0; i < scoreDocs.size() - 1; i++) {
+                ScoreDoc scoreDoc = scoreDocs.get(i);
+                int j = i + 1;
+                for (; j < scoreDocs.size(); j++) {
+                    ScoreDoc otherScoreDoc = scoreDocs.get(j);
+                    if (otherScoreDoc.doc != scoreDoc.doc) {
+                        break;
+                    }
+                    scoreDoc.score += otherScoreDoc.score;
+                }
+                if (j > i + 1) {
+                    scoreDocs.subList(i + 1, j).clear();
                 }
-                scoreDoc.score += otherScoreDoc.score;
-            }
-            if (j > i + 1) {
-                scoreDocs.subList(i + 1, j).clear();
             }
-        }
-        KnnScoreDocQueryBuilder knnQuery = new KnnScoreDocQueryBuilder(scoreDocs.toArray(new ScoreDoc[0]));
 
-        SearchSourceBuilder newSource = source.shallowCopy().knnSearch(List.of());
-        if (source.query() == null) {
-            newSource.query(knnQuery);
+            KnnScoreDocQueryBuilder knnQuery = new KnnScoreDocQueryBuilder(scoreDocs.toArray(new ScoreDoc[0]));
+            SearchSourceBuilder newSource = source.shallowCopy().knnSearch(List.of());
+            if (source.query() == null) {
+                newSource.query(knnQuery);
+            } else {
+                newSource.query(new BoolQueryBuilder().should(knnQuery).should(source.query()));
+            }
+            request.source(newSource);
         } else {
-            newSource.query(new BoolQueryBuilder().should(knnQuery).should(source.query()));
+            // this path will keep knn queries separate for ranking per shard
+            // if there are multiple knn queries
+
+            List<QueryBuilder> rankQueryBuilders = new ArrayList<>();
+            if (source.query() != null) {
+                rankQueryBuilders.add(source.query());
+            }
+
+            for (DfsKnnResults dfsKnnResults : knnResults) {
+                List<ScoreDoc> scoreDocs = new ArrayList<>();
+                for (ScoreDoc scoreDoc : dfsKnnResults.scoreDocs()) {
+                    if (scoreDoc.shardIndex == request.shardRequestIndex()) {
+                        scoreDocs.add(scoreDoc);
+                    }
+                }
+                scoreDocs.sort(Comparator.comparingInt(scoreDoc -> scoreDoc.doc));
+                KnnScoreDocQueryBuilder knnQuery = new KnnScoreDocQueryBuilder(scoreDocs.toArray(new ScoreDoc[0]));
+                rankQueryBuilders.add(knnQuery);
+            }
+
+            BoolQueryBuilder searchQuery = new BoolQueryBuilder();
+            for (QueryBuilder queryBuilder : rankQueryBuilders) {
+                searchQuery.should(queryBuilder);
+            }
```

Comment:
good question we do that exclusively for aggs as they are run against the conjunction of all the matches.

Reply:
We use this as the primary query for executing aggregations or hit tracking if those features are required. We may be able to figure out when we don't need to do this, but since a `null` gets rewritten to a `match_all` I think there may be some more work required there.